feat: CBOM support

san-zrl opened the pull request:
### Description

Enhances DT to read, persist, serve and export CBOM 1.6 data. See additional_details section for more information on what has been changed in particular. Note that there is a corresponding PR for hyades-frontend that enhances the UI to render CBOM data.

### Addressed Issue

[Issue #1538](https://github.com/DependencyTrack/hyades/issues/1538)

### Additional Details

- Added model classes and metric for CBOM data to `org.depencytrack.model`
- Enhanced ModelConverter to populate the additional model classes
- ModelConverter became quite massive so I split it into ModelConverter (cyclonedx -> internal model) and ModelExporter (internal model -> cyclonedx
- Added `org.depencytrack.persistence.v1.CryptoAssetsResource` with endpoints that serve the UI
- Enhanced exporter to generate CBOM data in download BOM 
- CBOM data are components of type `Classifier.CRYPTOGRAPHIC_ASSET` with CryptoProperties  and Occurrence attributes 

### Checklist

- [x] I have read and understand the [contributing guidelines]
- [ ] This PR fixes a defect, and I have provided tests to verify that the fix is effective
- [x] This PR implements an enhancement, and I have provided tests to verify that it works as intended
- [x] This PR introduces changes to the database model, and I have updated the [migration changelog] accordingly
- [ ] This PR introduces new or alters existing behavior, and I have updated the [documentation] accordingly

[contributing guidelines]: ../CONTRIBUTING.md#pull-requests
[documentation]: https://dependencytrack.github.io/hyades/latest/development/documentation/
[migration changelog]: https://dependencytrack.github.io/hyades/latest/development/database-migrations/

san-zrl commented:
Hi @VinodAnandan  - I looked into the tests and fixed the main problems particularly those that a related to the migration to cyclonedx 1..6. Not sure if the entire test set works because I never managed to run it successfully on my local system. I'm on a Mac M1 and getting testcontainers to run was a challenge. Main problems were

- There are some tests that rely on the jdo refresh feature to sync java objects with background changes in the database. One such example is [informWithExistingComponentPropertiesAndBomWithoutComponentProperties](https://github.com/DependencyTrack/hyades-apiserver/blob/b0581ffaa86bba08701ff9e69668763abfd4e21a/src/test/java/org/dependencytrack/tasks/BomUploadProcessingTaskTest.java#L1333). I couldn't get those to run and commented them out for the time being.
- [MigrationInitializerTest](https://github.com/DependencyTrack/hyades-apiserver/blob/b0581ffaa86bba08701ff9e69668763abfd4e21a/src/test/java/org/dependencytrack/persistence/migration/MigrationInitializerTest.java#L37) creates a temporary postgres container as the migration target. Some of the 4 tests in this class always fail. Behavior is pretty random.

nscuro commented:
> Not sure if the entire test set works because I never managed to run it successfully on my local system. I'm on a Mac M1 and getting testcontainers to run was a challenge

Can you elaborate what about the test containers was problematic? The team so far has been working predominantly with M1 macs, so that should not be a problem.

> There are some tests that rely on the jdo refresh feature to sync java objects with background changes in the database. [...] I couldn't get those to run and commented them out for the time being.

Can you share the errors you were getting here? Really all the refresh is doing is reloading the object from the database, so it's not much different from doing a `SELECT` query.

> [MigrationInitializerTest](https://github.com/DependencyTrack/hyades-apiserver/blob/b0581ffaa86bba08701ff9e69668763abfd4e21a/src/test/java/org/dependencytrack/persistence/migration/MigrationInitializerTest.java#L37) creates a temporary postgres container as the migration target. Some of the 4 tests in this class always fail. Behavior is pretty random.

Same as above, can you share the errors you're getting?

nscuro commented:
Also, have you tried if launching the API server with Dev Services works for you? https://dependencytrack.github.io/hyades/0.6.0-SNAPSHOT/development/testing/#api-server

san-zrl commented:
1. `BomUploadProcessingTaskTest#informWithExistingComponentPropertiesAndBomWithComponentProperties` (test report: [org.dependencytrack.tasks.BomUploadProcessingTaskTest.txt](https://github.com/user-attachments/files/17203222/org.dependencytrack.tasks.BomUploadProcessingTaskTest.txt)) fails with 
```
javax.jdo.JDOObjectNotFoundException: Object with id "org.dependencytrack.model.Component:33590 not found
```
2. Some test in `MigrationInitializerTest` (test report: [org.dependencytrack.persistence.migration.MigrationInitializerTest.txt](https://github.com/user-attachments/files/17203181/org.dependencytrack.persistence.migration.MigrationInitializerTest.txt)) always fail with the msg below. Some seem to work even though the the same mechanism underneath.
```
Caused by: org.postgresql.util.PSQLException: Connection to localhost:33271 refused. Check that the hostname and port are correct and that the postmaster is accepting TCP/IP connections.
```

nscuro commented:
I can reproduce the failures of the component property tests in this PR, but not in `main`. The tests expect a component to be modified, but it gets deleted instead. Which means DT couldn't match it's identity upon the second BOM upload.

The problem seems to be the modified `buildExactComponentIdentityQuery` method, where you currently have:

```java
        if (cid.getOid() != null) {
            filterParts.add("(cryptoAssetProperties != null && cryptoAssetProperties.oid == :oid)");
            params.put("oid", cid.getOid());
        } else {
            filterParts.add("cryptoAssetProperties != null && cryptoAssetProperties.oid == null");
        }
```

But it should be this instead:

```java
        if (cid.getOid() != null) {
            filterParts.add("(cryptoAssetProperties != null && cryptoAssetProperties.oid == :oid)");
            params.put("oid", cid.getOid());
        } else {
            filterParts.add("(cryptoAssetProperties == null || cryptoAssetProperties.oid == null)");
        }
```

The `MigrationInitializerTest` successfully completes for me. They also don't fail in CI. I wonder if perhaps there's something special about your local Docker setup? Are you using Docker or some other derivative such as Podman or Rancher?

san-zrl commented:
@nscuro: Good catch, thanks!

I'm using Rancher.

san-zrl commented:
@nscuro - The `buildExactComponentIdentityQuery` fix resolved most of the issues related to refresh(). However, there are still 3 `testUpdateMetricsUnchanged` tests that fail with assert violations on timestamps. Here's an example: 
[org.dependencytrack.tasks.metrics.ProjectMetricsUpdateTaskTest.txt](https://github.com/user-attachments/files/17207735/org.dependencytrack.tasks.metrics.ProjectMetricsUpdateTaskTest.txt). Could you have a look into this?

nscuro commented:
Will have a look at the failing tests.

Regarding Testcontainers, could this be relevant? https://docs.rancherdesktop.io/how-to-guides/using-testcontainers/#prerequisites

san-zrl commented:
> Regarding Testcontainers, could this be relevant? https://docs.rancherdesktop.io/how-to-guides/using-> testcontainers/#prerequisites

I've seen this. Rancher uses admin rights. Kubernetes is disabled, VM type is set to QEMU. My env settings are
```
DOCKER_HOST=unix:///Users/san/.rd/docker.sock
TESTCONTAINERS_DOCKER_SOCKET_OVERRIDE=/var/run/docker.sock
TESTCONTAINERS_HOST_OVERRIDE=localhost
TESTCONTAINERS_RYUK_DISABLED=true
```

nscuro commented:
> TESTCONTAINERS_RYUK_DISABLED=true

Not entirely sure, but maybe Ryuk not being there could be a problem. Can you try enabling it?

san-zrl commented:
TESTCONTAINERS_RYUK_DISABLED=false makes no difference for `MigrationInitializerTest`.

san-zrl commented:
Hi @nscuro, I pushed some code to bump the test coverage of the PR above the required threshold. The corresponding pipeline actions have been hanging since yesterday morning (https://github.com/DependencyTrack/hyades-apiserver/actions/runs/11177278805) with a message "This workflow is awaiting approval from a maintainer in https://github.com/DependencyTrack/hyades-apiserver/pull/933". Could you check what's going on?

nscuro commented:
@san-zrl PRs from first-time contributors need explicit approval for workflows to run. Approved it now.

san-zrl commented:
@VinodAnandan, thanks for letting me know. Diff test coverage is not quite on target Do you want me to further increase this number?

nscuro commented:
@san-zrl Don't sweat about the coverage too much for now. I haven't yet had the chance to thoroughly review the PR, and I am taking a few days "off" at the moment so there's limited time I'm spending on GitHub.

Assigning this to me, I'll get back to you ASAP.

nscuro commented on src/main/java/org/dependencytrack/persistence/ComponentQueryManager.java:
```diff
@@ -135,7 +135,7 @@ public List<Component> getAllComponents() {
     public List<Component> getAllComponents(Project project) {
         final Query<Component> query = pm.newQuery(Component.class, "project == :project");
-        query.getFetchPlan().setMaxFetchDepth(2);
+        query.getFetchPlan().setMaxFetchDepth(3);
```
What data are we trying to capture with this increase in fetch depth? I wonder if [`FetchGroup`](https://www.datanucleus.org/products/accessplatform_6_0/jdo/persistence.html#fetch_groups)s would be a better tool to avoid over-fetching of data that is not strictly needed.

san-zrl replied:
The frontend PR contains code for rendering crypto properties. They form an additional layer in the persistence model which must be fetched for rendering (fetch depth 2 was not enough). To keep things simple I just raised the fetch depth to make this work. We could define a dedicated crypto component fetch group and use this only when crypto assets are queried.

nscuro replied:
OK, I'd propose we deal with this once everything else in place, as it's just an optimization and DataNucleus' fetch behavior can be quite fiddly to tune.

nscuro commented on src/main/java/org/dependencytrack/model/Occurrence.java:
```diff
@@ -0,0 +1,114 @@
+    @JsonIgnore
+    private long id;
+
+    @Persistent
+    @Column(name = "BOM_REF", jdbcType = "VARCHAR", length = 64)
+    private String bomRef;
```
Do we need to store the BOM ref?

san-zrl replied:
No, we could drop this. The PR currently adds all crypto-related data fields that our scanner (https://github.com/IBM/cbomkit) generates. We can drop OCCURRENCES because the location of crypto assets in source is probably not important in DT.

nscuro reviewed (commented):
Some initial feedback with mostly questions from my side. Apologies for letting this sit for so long.

san-zrl commented on src/main/resources/migration/changelog-v5.6.0.xml:
```diff
@@ -107,4 +107,202 @@
+            <column autoIncrement="true" name="ID" type="BIGINT">
+                <constraints nullable="false" primaryKey="true" primaryKeyName="OCCURRENCES_PK"/>
+            </column>
+            <column name="BOM_REF" type="VARCHAR(64)"/>
+            <column name="LOCATION" type="VARCHAR(255)"/>
+            <column name="LINE" type="INT"/>
+            <column name="OFFSET" type="INT"/>
+            <column name="SYMBOL" type="INT"/>
+            <column name="ADDITIONAL_CONTEXT" type="VARCHAR(255)"/>
```
Hi @nscuro, thanks a lot for your review.

The code above comes from the definition of the OCCURRENCES table.  The PR currently adds all crypto-related data fields that our scanner (https://github.com/IBM/cbomkit) generates. We can drop OCCURRENCES because the location of crypto assets in source is probably not important in DT.

Nevertheless, the your point re. the size of the varchar columns is still valid. There is no proper reasoning behind the current numbers. We started with some guesstimates based on standard (https://cyclonedx.org/docs/1.6/json/) and examples (https://github.com/CycloneDX/bom-examples/tree/master/CBOM) and simply doubled if the initial numbers weren't sufficient. I'd gladly use other numbers with better foundation. If data size is an issue we should go through all varchars one-by-one and decide (a) does it have to be persisted and (b) if so, what size is needed.

nscuro replied:
> We can drop OCCURRENCES because the location of crypto assets in source is probably not important in DT.

Occurrence will likely come to DT in the future to cater to other use cases, but if it's not crucial for the crypto functionality it's good to drop here IMO.

> I'd gladly use other numbers with better foundation. If data size is an issue we should go through all varchars one-by-one and decide (a) does it have to be persisted and (b) if so, what size is needed.

In PostgreSQL it's all `TEXT` (unlimited length) behind the scenes, with an implicit `CHECK` constraint for the length. So other than preventing users from storing GBs of useless data, size constraints don't matter. With this in mind, I think choosing an initial length of `1024` should be fine.

san-zrl replied:
> Occurrence will likely come to DT in the future to cater to other use cases, but if it's not crucial for the crypto functionality it's good to drop here IMO
>....
> With this in mind, I think choosing an initial length of 1024 should be fine.



Okay, thanks.
1. I will remove Occurrence with the next commit.
2. VARCHAR length will default to 1024 unless we know better.

san-zrl commented on src/main/resources/migration/changelog-v5.6.0.xml:
```diff
@@ -107,4 +107,202 @@
+    </changeSet>
+
+    <changeSet id="v5.6.0-8" author="san-zrl">
+        <addColumn tableName="COMPONENT">
+            <column name="CRYPTO_PROPERTIES_ID" type="BIGINT"/>
+        </addColumn>
+
+        <createTable tableName="CRYPTO_PROPERTIES">
+            <column autoIncrement="true" name="ID" type="BIGINT">
+                <constraints nullable="false" primaryKey="true" primaryKeyName="CRYPTO_PROPERTIES_PK"/>
+            </column>
```
Yes, true. This would be a better idea but I didn't manage to make this work.

I tried to specify a COMPONENT_ID in CRYPTO_PROPERTIES that refers upwards to the enclosing component along with an FK constraint that would cause the right order of cascading deletes. This led to a problem when trying to persist the modified crypto asset properties object. I reproduced the error msg with stacktrace and appended it below for reference. It turns out that CryptoAssetProperties@5af49362 **is not the instance from ModelConverter** (in which the component id was properly set) but another instance created somewhere deep down in the transaction code.

```
2024-10-21 15:22:19,384 ERROR [Persist] Insert of object "org.dependencytrack.model.CryptoAssetProperties@5af49362" using statement "INSERT INTO "CRYPTO_PROPERTIES" ("ALGORITHM_PROPERTIES_ID","ASSET_TYPE","CERTIFICATE_PROPERTIES_ID","COMPONENT_ID","OID","PROTOCOL_PROPERTIES_ID","RELATED_MATERIAL_PROPERTIES_ID") VALUES (?,?,?,?,?,?,?)" failed : ERROR: null value in column "COMPONENT_ID" of relation "CRYPTO_PROPERTIES" violates not-null constraint
  Detail: Failing row contains (3, null, ALGORITHM, 3, null, null, null, 2.16.840.1.101.3.4.1.6). [bomSerialNumber=e8c355aa-2142-4084-a8c7-6d42c8610ba2, bomFormat=CycloneDX, bomUploadToken=6cb46db7-e009-4146-8615-122179457b63, projectName=x, bomSpecVersion=1.6, projectUuid=416f54cc-8621-492f-838f-b87ef3de9dad, projectVersion=null, bomVersion=1]
javax.jdo.JDODataStoreException: Insert of object "org.dependencytrack.model.CryptoAssetProperties@5af49362" using statement "INSERT INTO "CRYPTO_PROPERTIES" ("ALGORITHM_PROPERTIES_ID","ASSET_TYPE","CERTIFICATE_PROPERTIES_ID","COMPONENT_ID","OID","PROTOCOL_PROPERTIES_ID","RELATED_MATERIAL_PROPERTIES_ID") VALUES (?,?,?,?,?,?,?)" failed : ERROR: null value in column "COMPONENT_ID" of relation "CRYPTO_PROPERTIES" violates not-null constraint
  Detail: Failing row contains (3, null, ALGORITHM, 3, null, null, null, 2.16.840.1.101.3.4.1.6).
	at org.datanucleus.api.jdo.JDOAdapter.getJDOExceptionForNucleusException(JDOAdapter.java:608)
	at org.datanucleus.api.jdo.JDOPersistenceManager.flush(JDOPersistenceManager.java:2057)
	at org.dependencytrack.tasks.BomUploadProcessingTask.processComponents(BomUploadProcessingTask.java:529)
...
```

nscuro replied:
Just to be sure, is this how you modeled the relationship in this case? https://www.datanucleus.org/products/accessplatform_6_0/jdo/mapping.html#one_many_fk_bi

san-zrl replied:
I wanted the 'inner' dependent object to be fetched when the 'outer' object is queried and I wanted the cascading delete to work in both ways. The code does that, so I think we can leave it as is.

The fetching works since the `inner` objects (e.g., `CryptoAssetProperties`) are modelled as persistent members of the `outer` (e.g., ` Component`) object. Inner objects cannot exist without their 'outer' container. Thus, cascading delete must work in both ways:
1. Deleting an 'outer' object must delete all 'inner' objects. This is guaranteed by the `dependent = "true"` flag in the `Persistent` annotation. Here is an example:  https://github.com/DependencyTrack/hyades-apiserver/blob/279450cf3af0bec1ff23e712a2168b5d3508250b/src/main/java/org/dependencytrack/model/Component.java#L394-L397
2. Deleting an 'inner'  object renders the outer objects invalid and they must be deleted. This is guaranteed by the foreign key constraints. Here is an example: https://github.com/DependencyTrack/hyades-apiserver/blob/279450cf3af0bec1ff23e712a2168b5d3508250b/src/main/resources/migration/changelog-v5.6.0.xml#L269-L272

san-zrl commented on src/test/java/org/dependencytrack/tasks/BomUploadProcessingTaskTest.java:
I will add some tests. Your observation is right, we rely on the ORM magic.

san-zrl commented on src/main/java/org/dependencytrack/model/CryptoAlgorithmProperties.java:
```diff
@@ -0,0 +1,179 @@
+
+@PersistenceCapable(table= "ALGORITHM_PROPERTIES")
+@JsonInclude(JsonInclude.Include.NON_NULL)
+public class CryptoAlgorithmProperties implements Serializable {
```
We're not quite convinced that algorithm properties should be immutable global objects we could pick from. There are two main reasons:
1) Recently many new crypto algorithms have been proposed (e.g. quantum-safe ones) and this development is still ongoing.
2) As per the cyclonedx standard, algorithm property objects contain parameters which can vary between instances.

nscuro replied:
That makes sense, thanks for clarifying.

san-zrl commented on src/main/java/org/dependencytrack/model/CryptoCertificateProperties.java:
```diff
@@ -0,0 +1,143 @@
+    @Column(name = "NOT_VALID_AFTER", jdbcType = "TIMESTAMP")
+    @JsonSerialize(using = Iso8601DateSerializer.class)
+    private Date notValidAfter;
+    @Persistent
+    @Column(name = "SIGNATURE_ALGORITHM_REF", jdbcType = "VARCHAR", length=64)
+    private String signatureAlgorithmRef;
+    @Persistent
+    @Column(name = "SUBJECT_PUBLIC_KEY_REF", jdbcType = "VARCHAR", length=64)
+    private String subjectPublicKeyRef;
```
You're right, they should be resolved. I haven't done this yet because the issue appears in many places where the standard has xxxRef data fields. In most cases, these are bomrefs that could be resolved to components. Most prominently, the bomref resolution question arises in the dependencies. Here DT currently stores a JSON structure which imho cannot not be the final solution. Hence I thought that bomref resolution ist still an open issue and left the refs plain for the time being.

san-zrl replied:
In principle it's the same tree structure. The bomrefs in the ref fields must obviously point to other components in the BOM otherwise they would be dangling. In a sense, the standard is "over-specified": You can have ref fields that point to components **and** you can have `dependency` objects that also state that one component depends on another. It is not clear if the specification of a `signatureAlgorithmRef` requires the specification of a dependency object.

I think our final goal should be to model `signatureAlgorithmRef` as 
```java
 private Component subjectPublicKeyRef;
```
that is set as the result of the bomref resolution. Maybe we can make these fields transient because the get re-created on the fly when the BOM is loaded.

san-zrl commented on src/main/resources/migration/changelog-v5.6.0.xml:
```diff
@@ -107,4 +107,202 @@
+            <column name="REF" type="VARCHAR(64)"/>
+        </createTable>
+
+        <createTable tableName="CRYPTO_REFS">
+            <column autoIncrement="true" name="ID" type="BIGINT">
+                <constraints nullable="false" primaryKey="true" primaryKeyName="CRYPTO_REF_PK"/>
+            </column>
+            <column name="PROTOCOL_PROPERTY_ID" type="BIGINT"/>
+            <column name="CRYPTO_REF" type="VARCHAR(64)"/>
+        </createTable>
```
We could use arrays here but CRYPTO_REFS should probably be dropped for the moment. The standard mentions them (https://cyclonedx.org/docs/1.6/json/#components_items_cryptoProperties_protocolProperties_cryptoRefArray) but they are missing in the corresponding xsd definition (https://cyclonedx.org/schema/bom-1.6.xsd). Thus, the BOM parser currently does not recognize them. We filed a PR which was accepted but the xsd has not been updated yet.

san-zrl commented on src/main/resources/migration/changelog-v5.6.0.xml:
```diff
@@ -107,4 +107,202 @@
+            <column name="VERSION" type="VARCHAR(16)"/>
+        </createTable>
+
+        <createTable tableName="CRYPTO_FUNCTIONS">
+            <column autoIncrement="true" name="ID" type="BIGINT">
+                <constraints nullable="false" primaryKey="true" primaryKeyName="CRYPTO_FUNCTIONS_PK"/>
+            </column>
+            <column name="ALGORITHM_PROPERTY_ID" type="BIGINT"/>
+            <column name="CRYPTO_FUNCTION" type="VARCHAR(32)"/>
+        </createTable>
```
Yes, absolutely. I avoided array column types since they are not supported by all db systems. But if we stick to postgres using array types would greatly simplify the table structure. There are a couple of similar structures in CBOM 1.6 where we could migrate to arrays.

san-zrl commented on src/main/resources/migration/changelog-v5.6.0.xml:
```diff
@@ -107,4 +107,202 @@
+            <column name="CRYPTO_REF" type="VARCHAR(64)"/>
+        </createTable>
+
+        <createTable tableName="COMPONENT_OCCURRENCES">
+            <column name="COMPONENT_ID" type="BIGINT"/>
+            <column name="OCCURRENCE_ID" type="BIGINT"/>
+        </createTable>
```
The PR currently adds crypto-related data fields that our scanner (https://github.com/IBM/cbomkit) generates. We can drop OCCURRENCES because the location of crypto assets in source is probably not important in DT.

san-zrl commented on src/main/java/org/dependencytrack/parser/cyclonedx/util/ModelConverter.java:
As I said, we could drop Occurrences. Given that each crypto component can have many occurrences this would already save a lot. Other than Occurrences, the PR adds the entire cryptoProperties realm from https://cyclonedx.org/docs/1.6/json/ to persistence. There are some things that may be of less importance in the context of DT (e.g., the implementation platform and execution env in algorithm properties). These are only enum values and the saving potential is limited. We could have a telco and go through the data fields one by one. Perhaps there's too much detail in the standard for DT's purposes.

nscuro commented on src/main/resources/migration/changelog-v5.6.0.xml:
```diff
@@ -107,4 +107,202 @@
+            <column name="NAME" type="VARCHAR(64)"/>
+        </createTable>
+
+        <createTable tableName="CIPHER_SUITE_ALGORITHM">
+            <column name="CIPHER_SUITE_ID" type="BIGINT"/>
+            <column name="ALGORITHM" type="VARCHAR(64)"/>
+        </createTable>
```
We're [sticking to PostgreSQL only](https://dependencytrack.github.io/hyades/0.5.0/getting-started/changes-over-v4/#architecture-operations), so feel free to make use of PostgreSQL-specific features as much as you like.

san-zrl replied:
Okay, I'll try to turn them into array columns in the next commit. This would also reduce the size of our table zoo.

san-zrl commented:
Hi @nscuro - I just pushed commit https://github.com/DependencyTrack/hyades-apiserver/pull/933/commits/850eab98d30d4b9a684dec1e849991f9705c396d that addresses some of the issues we discussed:
- Not processing of occurrence data anymore. 
- Added some tests for ingesting Crypto BOMs.
- Collections of simple types are persisted as `text[]` columns in postgres. Join tables are only used for storing collections of complex types.

Issues not addressed yet:
- BOM refs are not yet resolved, still stored in plain.
- Crypto fields are still part of the default fetch group. No dedicated fetch plan yet.

nscuro commented on src/main/resources/migration/changelog-v5.6.0.xml:
```diff
@@ -107,4 +107,166 @@
+            <column name="REFS" type="TEXT[]"/>
+        </createTable>
+
+        <addForeignKeyConstraint baseTableName="COMPONENT" baseColumnNames="CRYPTO_PROPERTIES_ID"
+                                 constraintName="COMPONENT_CRYPTO_PROPERTIES_FK" deferrable="true" initiallyDeferred="true"
+                                 referencedTableName="CRYPTO_PROPERTIES" referencedColumnNames="ID"
+                                 onDelete="CASCADE" onUpdate="NO ACTION" validate="true"/>
+        <addForeignKeyConstraint baseTableName="CRYPTO_PROPERTIES" baseColumnNames="ALGORITHM_PROPERTIES_ID"
+                                 constraintName="CRYPTO_PROPERTIES_ALGORITHM_FK" deferrable="true" initiallyDeferred="true"
+                                 referencedTableName="ALGORITHM_PROPERTIES" referencedColumnNames="ID"
+                                 onDelete="CASCADE" onUpdate="NO ACTION" validate="true"/>
+        <addForeignKeyConstraint baseTableName="CRYPTO_PROPERTIES" baseColumnNames="CERTIFICATE_PROPERTIES_ID"
+                                 constraintName="CRYPTO_PROPERTIES_CERTIFICATE_FK" deferrable="true" initiallyDeferred="true"
+                                 referencedTableName="CERTIFICATE_PROPERTIES" referencedColumnNames="ID"
+                                 onDelete="CASCADE" onUpdate="NO ACTION" validate="true"/>
+        <addForeignKeyConstraint baseTableName="CRYPTO_PROPERTIES" baseColumnNames="RELATED_MATERIAL_PROPERTIES_ID"
+                                 constraintName="CRYPTO_PROPERTIES_RELATED_MATERIAL_FK" deferrable="true" initiallyDeferred="true"
+                                 referencedTableName="RELATED_CRYPTO_MATERIAL_PROPERTIES" referencedColumnNames="ID"
+                                 onDelete="CASCADE" onUpdate="NO ACTION" validate="true"/>
+        <addForeignKeyConstraint baseTableName="CRYPTO_PROPERTIES" baseColumnNames="PROTOCOL_PROPERTIES_ID"
+                                 constraintName="CRYPTO_PROPERTIES_PROTOCOL_FK" deferrable="true" initiallyDeferred="true"
+                                 referencedTableName="PROTOCOL_PROPERTIES" referencedColumnNames="ID"
+                                 onDelete="CASCADE" onUpdate="NO ACTION" validate="true"/>
```
Just a note so we won't forget: Postgres doesn't create indexes on FKs automatically. We'll need to create indexes explicitly if we need them to support fast lookups.

san-zrl commented on src/main/java/org/dependencytrack/model/Component.java:
Hi @nscuro - Thanks a lot for your comments and sorry for the late response. I'm currently caught up in some other projects so it may take a little while before I will come back with some code. Without having a deep insight yet, here is my preliminary answer on your thoughts.

> The issue with storing refs in DT is that they have no real meaning and could be totally random (e.g. UUIDs). So there's no real benefit of retaining them if they can't be resolved.

Agree. As a first step, I will change the ref fields to Component references (for the time being - see below) and add  a post-processing step to resolve the refs when the BOM has been loaded.

> IMO keeping Component and, say, CryptoComponent separate would make a lot of things more clear, and it would also perform better. 

Makes sense. The [CycloneDX Reference](https://cyclonedx.org/docs/1.6/json/) specifies components similarly to a union with a type selector. Except the name, components of type `cryptographic-asset` have not much in common with other component types. Their characteristics are specified in `cryptoProperties` which don't exist for other types. I'll look into this as a 2nd step.
